app/src/main/java/org/cru/godtools/ui/dashboard/tools/ToolsFragment.kt

```diff
@@ -2,24 +2,101 @@ package org.cru.godtools.ui.dashboard.tools
 
 import android.os.Bundle
 import androidx.fragment.app.viewModels
+import androidx.recyclerview.widget.ConcatAdapter
+import androidx.recyclerview.widget.RecyclerView
 import dagger.hilt.android.AndroidEntryPoint
+import javax.inject.Inject
+import org.ccci.gto.android.common.androidx.recyclerview.widget.addLayout
 import org.ccci.gto.android.common.sync.swiperefreshlayout.widget.SwipeRefreshSyncHelper
 import org.cru.godtools.R
 import org.cru.godtools.databinding.DashboardToolsFragmentBinding
+import org.cru.godtools.download.manager.GodToolsDownloadManager
 import org.cru.godtools.fragment.BasePlatformFragment
+import org.cru.godtools.model.Tool
+import org.cru.godtools.model.Translation
+import org.cru.godtools.ui.tooldetails.startToolDetailsActivity
+import org.cru.godtools.ui.tools.ToolsAdapter
+import org.cru.godtools.ui.tools.ToolsAdapterCallbacks
+import org.cru.godtools.ui.tools.ToolsAdapterViewModel
 
 @AndroidEntryPoint
-class ToolsFragment : BasePlatformFragment<DashboardToolsFragmentBinding>(R.layout.dashboard_tools_fragment) {
+class ToolsFragment :
+    BasePlatformFragment<DashboardToolsFragmentBinding>(R.layout.dashboard_tools_fragment),
+    ToolCategoriesAdapter.Callbacks,
+    ToolsAdapterCallbacks {
+    @Inject
+    internal lateinit var downloadManager: GodToolsDownloadManager
 
-    private val dataModel: ToolsFragmentDataModel by viewModels()
+    // region Data Model
+    private val dataModel: ToolsCategoryDataModel by viewModels()
+    private val toolsDataModel: ToolsAdapterViewModel by viewModels()
+    // endregion Data Model
+
+    // region Lifecycle
     override fun onBindingCreated(binding: DashboardToolsFragmentBinding, savedInstanceState: Bundle?) {
         super.onBindingCreated(binding, savedInstanceState)
         binding.refresh.setupSwipeRefresh()
-        binding.hasSpotlight = dataModel.hasSpotlight
+        binding.recyclerView.setupRecyclerView()
     }
 
     override fun onSyncData(helper: SwipeRefreshSyncHelper, force: Boolean) {
         super.onSyncData(helper, force)
         helper.sync(syncService.syncTools(force))
     }
+    //endregion Lifecycle
+
+    // region UI
+    private fun RecyclerView.setupRecyclerView() {
+        adapter = ConcatAdapter().apply {
```

Comment:
the `ConcatAdapter` should be configured to use stable ids

Reply:
@frett every time I add `.setHasStableIds(true)` the app crashes with the error illustrated.

![image](https://user-images.githubusercontent.com/970048/155209041-c66cc508-995e-46ec-8559-7505ab9114d1.png)

Reply:
you need to specify a config object when creating the `ConcatAdapter` to enable stable ids